File: saga/debugger.py
from __future__ import annotations

import json
import time
import tracemalloc
from collections import defaultdict
from collections.abc import Callable, Iterable
from pathlib import Path
from typing import Any

from .api import compile_file
from .interpreter import BuiltinFunction, Environment, Interpreter, SagaClass, UserFunction
from .native import Capabilities, NativeModule
from .tokens import Token


def _environment_snapshot(interpreter: Interpreter, env: Environment) -> dict[str, str]:
    # Walk lexical parents so a breakpoint inside a nested block can still
    # inspect the function/module variables that are actually in scope. Inner
    # bindings win, matching normal Saga name lookup.
    chain: list[Environment] = []
    current: Environment | None = env
    while current is not None:
        chain.append(current)
        current = current.parent
    visible: dict[str, object] = {}
    for scope in reversed(chain):
        for name, cell in scope.values.items():
            visible[name] = cell.value
    values: dict[str, str] = {}
    for name in sorted(visible):
        value = visible[name]
        if isinstance(value, (BuiltinFunction, NativeModule, SagaClass, UserFunction)):
            continue
        values[name] = interpreter.format_value(value)
    return values


def _environment_summary(interpreter: Interpreter, env: Environment) -> str:
    values = _environment_snapshot(interpreter, env)
    return "{" + ", ".join(f"{name}={value}" for name, value in values.items()) + "}"


def _write_json(path: str | Path, payload: Any) -> Path:
    target = Path(path).expanduser().resolve()
    target.parent.mkdir(parents=True, exist_ok=True)
    tmp = target.with_name(f".{target.name}.tmp")
    tmp.write_text(json.dumps(payload, ensure_ascii=False, indent=2, sort_keys=True) + "\n", encoding="utf-8")
    tmp.replace(target)
    return target


def debug_file(
    path: str | Path,
    *,
    trace: bool = False,
    breakpoints: Iterable[int] = (),
    watches: Iterable[str] = (),
    record_path: str | Path | None = None,
    max_events: int = 100_000,
    output: Callable[[str], None] = print,
    debug_output: Callable[[str], None] = print,
    precision: int = 50,
    step_limit: int | None = None,
    capabilities: Capabilities | None = None,
) -> dict[str, Any]:
    """Run a checked Saga program with deterministic statement-level debugging.

    The debugger remains non-interactive so it is reproducible in terminals, CI
    and editors. 0.37 adds watch values and a bounded JSON execution recording;
    the recording can be inspected backwards to reproduce state at a statement
    boundary without mutating the debugged process.
    """
    entry_input = Path(path).expanduser()
    loaded = compile_file(str(entry_input))
    entry = loaded.entry
    selected = {int(line) for line in breakpoints}
    watched = tuple(dict.fromkeys(str(name) for name in watches if str(name)))
    if any(line < 1 for line in selected):
        raise ValueError("breakpoint lines must be >= 1")
    if max_events < 1:
        raise ValueError("max_events must be >= 1")

    interpreter: Interpreter
    events: list[dict[str, Any]] = []
    dropped = 0
    sequence = 0

    def hook(token: Token, env: Environment) -> None:
        nonlocal dropped, sequence
        sequence += 1
        snapshot = _environment_snapshot(interpreter, env)
        is_break = token.line in selected
        if trace or is_break:
            label = "break" if is_break else "trace"
            filename = token.filename or str(entry)
            watch_text = ""
            if watched:
                selected_values = ", ".join(f"{name}={snapshot.get(name, '<unbound>')}" for name in watched)
                watch_text = f" watches={{{selected_values}}}"
            debug_output(
                f"[{label}] {filename}:{token.line}:{token.column} "
                f"locals=" + "{" + ", ".join(f"{k}={v}" for k, v in snapshot.items()) + "}" + watch_text
            )
        if record_path is not None:
            if len(events) < max_events:
                events.append({
                    "seq": sequence,
                    "file": token.filename or str(entry),
                    "line": token.line,
                    "column": token.column,
                    "breakpoint": is_break,
                    "locals": snapshot,
                    "watches": {name: snapshot.get(name) for name in watched},
                    "watch": {name: snapshot.get(name, "<unbound>") for name in watched},
                })
            else:
                dropped += 1

    interpreter = Interpreter(
        str(entry), output=output, precision=precision, step_limit=step_limit,
        capabilities=capabilities, debug_hook=hook,
    )
    try:
        interpreter.interpret(loaded.program)
    finally:
        interpreter.close()

    report = {
        "schema": "saga.debug-record.v1",
        "source": str(entry),
        "events_recorded": len(events),
        "events_dropped": dropped,
        "truncated": dropped > 0,
        "breakpoints": sorted(selected),
        "watches": list(watched),
        "events": events,
    }
    if record_path is not None:
        _write_json(record_path, report)
    return report


def profile_file(
    path: str | Path,
    *,
    output: Callable[[str], None] = lambda _text: None,
    precision: int = 50,
    step_limit: int | None = None,
    capabilities: Capabilities | None = None,
    report_path: str | Path | None = None,
    top: int = 20,
) -> dict[str, Any]:
    """Profile statement-boundary execution time, hit counts and Python heap peak.

    Time between two statement hooks is attributed to the previous source
    location. This produces a low-overhead line/statement profile suitable for
    finding hot control paths. It is deliberately described as interval time,
    not CPU-instruction attribution.
    """
    if top < 1:
        raise ValueError("top must be >= 1")
    entry_input = Path(path).expanduser()
    loaded = compile_file(str(entry_input))
    entry = loaded.entry
    hits: dict[tuple[str, int, int], int] = defaultdict(int)
    interval_ns: dict[tuple[str, int, int], int] = defaultdict(int)
    previous_key: tuple[str, int, int] | None = None
    previous_ns = time.perf_counter_ns()
    interpreter: Interpreter

    def hook(token: Token, _env: Environment) -> None:
        nonlocal previous_key, previous_ns
        now = time.perf_counter_ns()
        if previous_key is not None:
            interval_ns[previous_key] += max(0, now - previous_ns)
        key = (token.filename or str(entry), token.line, token.column)
        hits[key] += 1
        previous_key = key
        previous_ns = now

    tracemalloc.start()
    started = time.perf_counter_ns()
    interpreter = Interpreter(
        str(entry), output=output, precision=precision, step_limit=step_limit,
        capabilities=capabilities, debug_hook=hook,
    )
    try:
        interpreter.interpret(loaded.program)
    finally:
        finished = time.perf_counter_ns()
        if previous_key is not None:
            interval_ns[previous_key] += max(0, finished - previous_ns)
        interpreter.close()
        current_bytes, peak_bytes = tracemalloc.get_traced_memory()
        tracemalloc.stop()

    rows = []
    for key in sorted(hits):
        file_name, line, column = key
        ns = interval_ns.get(key, 0)
        rows.append({
            "file": file_name,
            "line": line,
            "column": column,
            "hits": hits[key],
            "interval_ns": ns,
            "avg_interval_ns": ns // hits[key] if hits[key] else 0,
        })
    rows.sort(key=lambda row: (-row["interval_ns"], row["file"], row["line"], row["column"]))
    report = {
        "schema": "saga.statement-profile.v1",
        "source": str(entry),
        "elapsed_ns": max(0, finished - started),
        "python_heap_current_bytes": current_bytes,
        "python_heap_peak_bytes": peak_bytes,
        "statement_events": sum(hits.values()),
        "locations": rows,
        "top": rows[:top],
        "timing_model": "time from one Saga statement hook to the next is attributed to the previous location",
    }
    if report_path is not None:
        _write_json(report_path, report)
    return report

File: implementations/go/cmd/saga-go/debugger.go
//go:build !sagaruntime

package main

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"runtime"
	"sort"
	"strconv"
	"strings"
	"time"
)

func debugEnvSnapshot(e *Env) map[string]string {
	out := map[string]string{}
	if e == nil {
		return out
	}
	chain := []*Env{}
	for current := e; current != nil; current = current.Parent {
		chain = append(chain, current)
	}
	// Outer scopes are applied first so inner lexical bindings shadow them.
	for i := len(chain) - 1; i >= 0; i-- {
		for name, cell := range chain[i].Values {
			if cell == nil {
				continue
			}
			switch cell.V.(type) {
			case *NativeFunc, CoreModule, *Class, *Function:
				continue
			}
			if cell.Moved {
				out[name] = "<moved>"
			} else {
				out[name] = formatValue(cell.V, false)
			}
		}
	}
	return out
}

func debugEnvSummary(e *Env) string {
	values := debugEnvSnapshot(e)
	keys := make([]string, 0, len(values))
	for k := range values {
		keys = append(keys, k)
	}
	sort.Strings(keys)
	parts := make([]string, 0, len(keys))
	for _, k := range keys {
		parts = append(parts, k+"="+values[k])
	}
	return "{" + strings.Join(parts, ", ") + "}"
}

func writeJSONAtomic(path string, payload any) error {
	if path == "" {
		return nil
	}
	abs, err := filepath.Abs(path)
	if err != nil {
		return err
	}
	if err = os.MkdirAll(filepath.Dir(abs), 0755); err != nil {
		return err
	}
	f, err := os.CreateTemp(filepath.Dir(abs), "."+filepath.Base(abs)+".*.tmp")
	if err != nil {
		return err
	}
	tmp := f.Name()
	committed := false
	defer func() {
		_ = f.Close()
		if !committed {
			_ = os.Remove(tmp)
		}
	}()
	enc := json.NewEncoder(f)
	enc.SetIndent("", "  ")
	enc.SetEscapeHTML(false)
	if err = enc.Encode(payload); err != nil {
		return err
	}
	if err = f.Sync(); err != nil {
		return err
	}
	if err = f.Close(); err != nil {
		return err
	}
	if err = atomicReplacePath(tmp, abs); err != nil {
		return err
	}
	committed = true
	return nil
}

func runDebugger(args []string) int {
	if len(args) < 1 {
		fmt.Fprintln(os.Stderr, "source file required")
		return 64
	}
	path := args[0]
	breaks := map[int]bool{}
	trace := false
	watches := []string{}
	recordPath := ""
	maxEvents := 100000
	for j := 1; j < len(args); j++ {
		switch args[j] {
		case "--trace":
			trace = true
		case "--break":
			if j+1 >= len(args) {
				fmt.Fprintln(os.Stderr, "--break requires line")
				return 64
			}
			n, e := strconv.Atoi(args[j+1])
			if e != nil || n < 1 {
				fmt.Fprintln(os.Stderr, "invalid breakpoint")
				return 64
			}
			breaks[n] = true
			j++
		case "--watch":
			if j+1 >= len(args) || strings.TrimSpace(args[j+1]) == "" {
				fmt.Fprintln(os.Stderr, "--watch requires name")
				return 64
			}
			watches = append(watches, args[j+1])
			j++
		case "--record":
			if j+1 >= len(args) {
				fmt.Fprintln(os.Stderr, "--record requires path")
				return 64
			}
			recordPath = args[j+1]
			j++
		case "--max-events":
			if j+1 >= len(args) {
				fmt.Fprintln(os.Stderr, "--max-events requires integer")
				return 64
			}
			n, e := strconv.Atoi(args[j+1])
			if e != nil || n < 1 || n > 1000000 {
				fmt.Fprintln(os.Stderr, "--max-events must be 1..1000000")
				return 64
			}
			maxEvents = n
			j++
		default:
			fmt.Fprintln(os.Stderr, "unknown debug option:", args[j])
			return 64
		}
	}
	stmts, err := loadProgram(path)
	if err != nil {
		return printDiagnostic(err)
	}
	c := NewChecker()
	if err = c.Check(stmts); err != nil {
		return printDiagnostic(err)
	}
	it := NewInterpreter(c, nil)
	events := []map[string]any{}
	truncated := false
	it.DebugHook = func(t Token, e *Env) {
		values := debugEnvSnapshot(e)
		isBreak := breaks[t.Line]
		if trace || isBreak {
			label := "trace"
			if isBreak {
				label = "break"
			}
			watchText := ""
			if len(watches) > 0 {
				parts := []string{}
				for _, name := range watches {
					value, ok := values[name]
					if !ok {
						value = "<unbound>"
					}
					parts = append(parts, name+"="+value)
				}
				watchText = " watch={" + strings.Join(parts, ", ") + "}"
			}
			fmt.Fprintf(os.Stderr, "[%s] %s:%d:%d locals=%s%s\n", label, t.File, t.Line, t.Col, debugEnvSummary(e), watchText)
		}
		if recordPath != "" {
			if len(events) < maxEvents {
				watched := map[string]string{}
				for _, name := range watches {
					value, ok := values[name]
					if !ok {
						value = "<unbound>"
					}
					watched[name] = value
				}
				events = append(events, map[string]any{
					"seq": len(events), "file": t.File, "line": t.Line, "column": t.Col,
					"breakpoint": isBreak, "locals": values, "watch": watched,
				})
			} else {
				truncated = true
			}
		}
	}
	if err = it.Interpret(stmts); err != nil {
		return printDiagnostic(err)
	}
	if recordPath != "" {
		report := map[string]any{
			"schema": "saga.debug-record.v1", "implementation": "saga-native-go",
			"source": path, "events": events, "event_count": len(events), "max_events": maxEvents,
			"truncated": truncated, "watches": watches,
		}
		if err = writeJSONAtomic(recordPath, report); err != nil {
			fmt.Fprintln(os.Stderr, "write debug record:", err)
			return 74
		}
	}
	return 0
}

type profileLocation struct {
	File     string `json:"file"`
	Line     int    `json:"line"`
	Column   int    `json:"column"`
	Hits     int64  `json:"hits"`
	Interval int64  `json:"interval_ns"`
}

func runProfiler(args []string) int {
	if len(args) < 1 {
		fmt.Fprintln(os.Stderr, "source file required")
		return 64
	}
	path := args[0]
	reportPath := ""
	top := 20
	for j := 1; j < len(args); j++ {
		switch args[j] {
		case "--json":
			if j+1 >= len(args) {
				fmt.Fprintln(os.Stderr, "--json requires path")
				return 64
			}
			reportPath = args[j+1]
			j++
		case "--top":
			if j+1 >= len(args) {
				fmt.Fprintln(os.Stderr, "--top requires integer")
				return 64
			}
			n, e := strconv.Atoi(args[j+1])
			if e != nil || n < 1 || n > 1000 {
				fmt.Fprintln(os.Stderr, "--top must be 1..1000")
				return 64
			}
			top = n
			j++
		default:
			fmt.Fprintln(os.Stderr, "unknown profile option:", args[j])
			return 64
		}
	}
	stmts, err := loadProgram(path)
	if err != nil {
		return printDiagnostic(err)
	}
	c := NewChecker()
	if err = c.Check(stmts); err != nil {
		return printDiagnostic(err)
	}
	it := NewInterpreter(c, nil)
	locations := map[string]*profileLocation{}
	var previous *profileLocation
	var previousAt time.Time
	var before, after runtime.MemStats
	runtime.ReadMemStats(&before)
	started := time.Now()
	it.DebugHook = func(t Token, _ *Env) {
		now := time.Now()
		if previous != nil && !previousAt.IsZero() {
			previous.Interval += now.Sub(previousAt).Nanoseconds()
		}
		key := fmt.Sprintf("%s\x00%d\x00%d", t.File, t.Line, t.Col)
		row := locations[key]
		if row == nil {
			row = &profileLocation{File: t.File, Line: t.Line, Column: t.Col}
			locations[key] = row
		}
		row.Hits++
		previous = row
		previousAt = now
	}
	if err = it.Interpret(stmts); err != nil {
		return printDiagnostic(err)
	}
	ended := time.Now()
	if previous != nil && !previousAt.IsZero() {
		previous.Interval += ended.Sub(previousAt).Nanoseconds()
	}
	runtime.ReadMemStats(&after)
	rows := make([]profileLocation, 0, len(locations))
	var events int64
	for _, row := range locations {
		rows = append(rows, *row)
		events += row.Hits
	}
	sort.Slice(rows, func(i, j int) bool {
		if rows[i].Interval != rows[j].Interval {
			return rows[i].Interval > rows[j].Interval
		}
		if rows[i].File != rows[j].File {
			return rows[i].File < rows[j].File
		}
		if rows[i].Line != rows[j].Line {
			return rows[i].Line < rows[j].Line
		}
		return rows[i].Column < rows[j].Column
	})
	topRows := rows
	if len(topRows) > top {
		topRows = topRows[:top]
	}
	for _, row := range topRows {
		fmt.Printf("%10.3f ms  hits=%6d  %s:%d:%d\n", float64(row.Interval)/1e6, row.Hits, row.File, row.Line, row.Column)
	}
	peakDelta := uint64(0)
	if after.HeapSys > before.HeapSys {
		peakDelta = after.HeapSys - before.HeapSys
	}
	report := map[string]any{
		"schema": "saga.statement-profile.v1", "implementation": "saga-native-go", "source": path,
		"timing_model": "elapsed interval between consecutive Saga statement hooks attributed to the preceding statement; not CPU instruction attribution",
		"elapsed_ns":   ended.Sub(started).Nanoseconds(), "statement_events": events, "locations": rows,
		"top": topRows, "go_heap_alloc_bytes": after.HeapAlloc, "go_heap_sys_growth_bytes": peakDelta,
	}
	if reportPath != "" {
		if err = writeJSONAtomic(reportPath, report); err != nil {
			fmt.Fprintln(os.Stderr, "write profile report:", err)
			return 74
		}
	}
	fmt.Printf("elapsed=%.3f ms heap_alloc=%d bytes\n", float64(ended.Sub(started).Nanoseconds())/1e6, after.HeapAlloc)
	return 0
}
